node/src/main/scala/com/wavesplatform/state/diffs/TransactionDiffer.scala

```diff
@@ -99,4 +105,54 @@ object TransactionDiffer {
       case stx: SponsorFeeTransaction        => AssetTransactionsDiff.sponsor(blockchain, currentBlockTimestamp)(stx)
       case _                                 => Left(UnsupportedTransactionType)
     }
+
+  private def balance(blockchain: Blockchain, tx: Transaction, diff: Diff): TracedResult[ValidationError, Diff] =
+    stats.balanceValidation
+      .measureForType(tx.typeId) {
+        TracedResult(BalanceDiffValidation(blockchain)(diff))
+      }
+
+  private def failedScriptTransactionDiff(blockchain: Blockchain, tx: Transaction): TracedResult[ValidationError, Diff] =
+    for {
+      portfolios <- (tx, tx.assetFee) match {
+        case (tx: ProvenTransaction, (Waves, fee)) => TracedResult(Right(Map(tx.sender.toAddress -> Portfolio(-fee, LeaseBalance.empty, Map.empty))))
+        case (tx: ProvenTransaction, (asset @ IssuedAsset(_), fee)) =>
+          TracedResult {
+            for {
+              assetInfo <- blockchain
+                .assetDescription(asset)
+                .toRight(GenericError(s"Asset $asset does not exist, cannot be used to pay fees"))
+              wavesFee <- Either.cond(
+                assetInfo.sponsorship > 0,
+                Sponsorship.toWaves(fee, assetInfo.sponsorship),
+                GenericError(s"Asset $asset is not sponsored, cannot be used to pay fees")
+              )
+            } yield {
+              Map(tx.sender.toAddress          -> Portfolio(0, LeaseBalance.empty, Map(asset         -> -fee))) |+|
+                Map(assetInfo.issuer.toAddress -> Portfolio(-wavesFee, LeaseBalance.empty, Map(asset -> fee)))
+            }
+          }
+        case _ => TracedResult(Left(TransactionValidationError(GenericError(s"Can't accept failed script on $tx"), tx)))
+      }
+      diff <- balance(blockchain, tx, Diff.failed(tx, portfolios))
+    } yield diff
+
+  private def acceptFailedScript(blockchain: Blockchain, tx: Transaction): Boolean =
+    (tx.typeId == InvokeScriptTransaction.typeId || tx.typeId == ExchangeTransaction.typeId) &&
+      blockchain.isFeatureActivated(BlockchainFeatures.AcceptFailedScriptTransaction)
+
+  private object isFailedScript {
+    def unapply(result: TracedResult[ValidationError, Diff]): Boolean =
+      result match {
+        case TracedResult(Left(TransactionValidationError(ScriptExecutionError(_, _, true), _)), _)       => true
+        case TracedResult(Left(TransactionValidationError(TransactionNotAllowedByScript(_, true), _)), _) => true
```

Comment:
How about adding a marker trait for script-related errors? Also `TransactionNotAllowedByScript` is ambiguous: which script exactly, asset script or account script? In this particular place this difference might be significant.